test/integration/job/job_test.go

```diff
@@ -0,0 +1,338 @@
+		t.Fatalf("Failed setting phase %s on Job Pods: %v", v1.PodFailed, err)
+	}
+	validateJobPodsStatus(ctx, t, clientSet, jobObj, podsByStatus{
+		Active: 5,
+		Failed: 2,
+	})
+	// Once one Pod succeeds, no more Pods are created, even if some fail.
+	if err := setJobPodsPhase(ctx, t, clientSet, jobObj, v1.PodSucceeded, 1); err != nil {
+		t.Fatalf("Failed setting phase %s on Job Pod: %v", v1.PodSucceeded, err)
+	}
+	validateJobPodsStatus(ctx, t, clientSet, jobObj, podsByStatus{
+		Failed:    2,
+		Succeeded: 1,
+		Active:    4,
+	})
+	if err := setJobPodsPhase(ctx, t, clientSet, jobObj, v1.PodFailed, 2); err != nil {
+		t.Fatalf("Failed setting phase %s on Job Pods: %v", v1.PodFailed, err)
+	}
+	validateJobPodsStatus(ctx, t, clientSet, jobObj, podsByStatus{
+		Failed:    4,
+		Succeeded: 1,
+		Active:    2,
+	})
+	// No more Pods are created after remaining Pods succeed.
+	if err := setJobPodsPhase(ctx, t, clientSet, jobObj, v1.PodSucceeded, 2); err != nil {
+		t.Fatalf("Failed setting phase %s on Job Pods: %v", v1.PodSucceeded, err)
+	}
+	validateJobSucceeded(ctx, t, clientSet, jobObj)
+	validateJobPodsStatus(ctx, t, clientSet, jobObj, podsByStatus{
+		Failed:    4,
+		Succeeded: 3,
+	})
+}
+
+func TestParallelJobWithCompletions(t *testing.T) {
+	closeFn, restConfig, clientSet, ns := setup(t, "simple-job")
+	defer closeFn()
+	ctx, cancel := startJobController(restConfig, clientSet)
+	defer cancel()
+
+	jobName := "parallel-job-completions"
+	jobObj, err := clientSet.BatchV1().Jobs(ns.Name).Create(ctx, &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: ns.Name,
+			Name:      jobName,
+		},
+		Spec: batchv1.JobSpec{
+			Parallelism: pointer.Int32Ptr(4),
+			Completions: pointer.Int32Ptr(6),
+			Template: v1.PodTemplateSpec{
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{
+						{Name: "foo", Image: "bar"},
+					},
+					RestartPolicy: v1.RestartPolicyNever,
+				},
+			},
+		},
+	}, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to create Job: %v", err)
+	}
+	validateJobPodsStatus(ctx, t, clientSet, jobObj, podsByStatus{
+		Active: 4,
+	})
+	// Failed Pods are replaced.
+	if err := setJobPodsPhase(ctx, t, clientSet, jobObj, v1.PodFailed, 2); err != nil {
+		t.Fatalf("Failed setting phase %s on Job Pods: %v", v1.PodFailed, err)
+	}
+	validateJobPodsStatus(ctx, t, clientSet, jobObj, podsByStatus{
+		Active: 4,
+		Failed: 2,
+	})
+	// Pods are created until the number of succeeded Pods equals completions.
+	if err := setJobPodsPhase(ctx, t, clientSet, jobObj, v1.PodSucceeded, 3); err != nil {
+		t.Fatalf("Failed setting phase %s on Job Pod: %v", v1.PodSucceeded, err)
+	}
+	validateJobPodsStatus(ctx, t, clientSet, jobObj, podsByStatus{
+		Failed:    2,
+		Succeeded: 3,
+		Active:    3,
+	})
+	// No more Pods are created after the Job completes.
+	if err := setJobPodsPhase(ctx, t, clientSet, jobObj, v1.PodSucceeded, 3); err != nil {
```

Comment:
can we test this: succeed one more pod (a=2, s=c=4, f=2), then fail one more (a=1, s=4, f=3), and then make sure new active pods aren't created (because c=4 achieved)?

Reply:
Note that we have parallelism=4 in this test, and we can see that no more Pods are created. Since no more Pods are created, we cannot fail any Pod.

But what you mention could happen in a queue-pattern, which is already covered in such test.

Reply:
>Since no more Pods are created, we cannot fail any Pod

That's true if we reach s=6, but I meant that instead of like 216 (succeeding the last three active pods at the same time), we should succeed just one pod (f=2, s=c=4, a=2), then fail another (f=3, s=c=4) and make sure a=1 even after failing a pod (i.e. the failed pod isn't replaced).

Reply:
succeeded needs to be at 6 in order for the failed pod not to be replaced.